File: gap-trap/templates/gates/instruction_gate_test.py
"""Instruction gate. Reference implementation from gap-trap (pytest).

Same assertions as instruction-gate.test.ts: every name a contract cites
exists where the contract says it lives, the portable core stays portable,
the always-loaded files stay small, cited commits exist, knowledge files hold
no private data, docs cite rule IDs that exist, and the grep gates settle the
contract Never clauses a text search can.

Every check also asserts its own denominator: how many tokens it resolved,
how many files it scanned. A check that quietly measures nothing reports the
same green as one that measured everything (M2).

ADAPT the config block; keep the assertions.
"""
from __future__ import annotations

import io
import re
import subprocess
import tokenize
from pathlib import Path

import pytest

# ---- config ---------------------------------------------------------------
REPO = Path(__file__).resolve().parents[1]          # ADAPT: repo root relative to this file
SRC_DIR = "src"                                     # ADAPT
SOURCE_EXT = (".py",)
TEST_DIRS = {"tests", "test", "__tests__"}
FORBIDDEN_IN_CORE = ["{{PRODUCT}}", "{{FRAMEWORK}}", SRC_DIR, "agents/project"]  # ADAPT
WORD_BUDGET = 4000                                  # ADAPT: current count plus room (C7)
MIN_CONTRACTS = 2  # the honest count; never pad
ALWAYS_LOADED = ["AGENTS.md", "AGENTS.project.md", "CLAUDE.md"]  # ADAPT: loaded every session
KNOWLEDGE_FILES = [
    "agents/project/domain-context.md",
    "agents/project/glossary.md",
    "agents/project/out-of-scope.md",
    "agents/generic/agent-workflows.md",
]
DOCS_DIR = "docs"                                   # ADAPT: "" skips the rule-ID check
# ADAPT: (name, pattern over comment-stripped source, exempt-path predicate)
GREP_GATES = [
    ("Logging: no print outside the logger", re.compile(r"\bprint\("), lambda rel: rel == "lib/logger.py"),
    ("HTTP: no raw requests/httpx outside the wrapper", re.compile(r"\b(requests|httpx)\.(get|post|put|delete|request)\("), lambda rel: rel.startswith("lib/http")),
]
# ---- end config -----------------------------------------------------------

SRC = REPO / SRC_DIR
FIELD_RE = re.compile(r"^(Owns|Path|Never|Gate):")


def read(p: Path) -> str:
    return p.read_text(encoding="utf-8")


def source_files() -> list[Path]:
    return [p for p in SRC.rglob("*") if p.suffix in SOURCE_EXT and "node_modules" not in p.parts]


def is_test_path(p: Path) -> bool:
    rel = p.relative_to(SRC)
    return any(part in TEST_DIRS for part in rel.parts) or p.name.startswith("test_") or p.name.endswith("_test.py")


def parse_contracts(md: str) -> list[tuple[str, str]]:
    section = md.split("## Architecture contracts", 1)[1].split("\n## ", 1)[0] if "## Architecture contracts" in md else ""
    out = []
    for block in section.split("\n### ")[1:]:
        name, _, body = block.partition("\n")
        out.append((name.strip(), body))
    return out


def fold_fields(body: str) -> list[str]:
    """Fold a contract body's wrapped lines back into the field they continue.

    Markdown prose wraps, and a ``Path:`` line long enough to wrap used to
    have everything past the wrap silently dropped from the check.
    """
    out: list[str] = []
    open_field = False
    for raw in body.splitlines():
        line = raw.strip()
        if not line:
            open_field = False
        elif FIELD_RE.match(line):
            out.append(line)
            open_field = True
        elif open_field:
            out[-1] += " " + line
    return out


def backtick_tokens(line: str) -> list[str]:
    return [t.removesuffix("()") for t in re.findall(r"`([^`]+)`", line)]


def text_under(rel: str) -> str:
    """The text under a repo-relative path, file or directory."""
    p = REPO / rel
    if not p.exists():
        return ""
    if p.is_dir():
        return "\n".join(read(f) for f in p.rglob("*") if f.suffix in SOURCE_EXT and f.is_file())
    return read(p)


CONTRACTS = parse_contracts(read(REPO / "AGENTS.project.md"))
# A contract symbol found only in a test file is not a symbol the code uses.
NON_TEST_SOURCE_TEXT = "\n".join(read(p) for p in source_files() if not is_test_path(p))


def test_contracts_have_all_four_lines():
    assert len(CONTRACTS) >= MIN_CONTRACTS
    for name, body in CONTRACTS:
        fields = fold_fields(body)
        for field in ("Owns:", "Path:", "Never:", "Gate:"):
            assert any(l.startswith(field) for l in fields), f"{name} missing {field}"


def test_every_symbol_and_path_in_contracts_exists():
    checked_overall = 0
    for name, body in CONTRACTS:
        lines = [l for l in fold_fields(body) if l.startswith(("Path:", "Gate:"))]
        checked = 0
        for line in lines:
            tokens = backtick_tokens(line)
            paths = [t for t in tokens if "/" in t]
            symbols = [t for t in tokens if "/" not in t]
            for token in paths:
                assert (REPO / token).exists(), f"{name}: path {token} missing"
                checked += 1
            # A Path line names where the subsystem lives, so its symbols are
            # looked for there. Resolving them anywhere in the tree let a
            # symbol that had moved out of its own module keep the contract
            # green.
            scoped = line.startswith("Path:") and paths
            haystack = "\n".join(text_under(p) for p in paths) if scoped else NON_TEST_SOURCE_TEXT
            where = ", ".join(paths) if scoped else f"{SRC_DIR} (non-test files)"
            for token in symbols:
                assert re.search(rf"\b{re.escape(token)}\b", haystack), f"{name}: symbol {token} not found in {where}"
                checked += 1
        assert checked > 0, f"{name}: its Path/Gate lines name nothing the gate can check"
        checked_overall += checked
    assert checked_overall > 0, "no contract token was checked at all"


def test_core_is_portable():
    core = read(REPO / "AGENTS.md").lower()
    for token in FORBIDDEN_IN_CORE:
        assert token.lower() not in core, f'AGENTS.md contains "{token}"'


def test_always_loaded_files_within_word_budget():
    # Every file has to be there. Summing whatever happens to exist means
    # deleting an always-loaded file reads as coming in under budget.
    for f in ALWAYS_LOADED:
        assert (REPO / f).exists(), f"{f} is missing; the budget would fall for free"
    total = sum(len(read(REPO / f).split()) for f in ALWAYS_LOADED)
    assert total <= WORD_BUDGET, f"combined {total} words > budget {WORD_BUDGET}"


def test_cited_commit_hashes_exist():
    md = read(REPO / "agents/project/domain-context.md")
    # 7 to 40 hex chars with a digit: --oneline prints 7, and the digit keeps
    # English words made of a-f (acceded, defaced) out.
    for h in sorted({h for h in re.findall(r"\b[0-9a-f]{7,40}\b", md) if re.search(r"\d", h)}):
        r = subprocess.run(["git", "cat-file", "-e", f"{h}^{{commit}}"], cwd=REPO, capture_output=True)
        assert r.returncode == 0, f"cited commit {h} not found in history"


def test_knowledge_files_hold_no_private_data():
    for f in KNOWLEDGE_FILES:
        p = REPO / f
        if not p.exists():
            continue
        text = read(p)
        assert not re.search(r"\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b", text), f"{f} contains an IP address"
        # The domain needs a letter TLD, or a pinned version such as
        # `pytest@1.2.3` reads as an email and the gate cries wolf.
        assert not re.search(r"\b[\w.+-]+@[\w-]+\.[A-Za-z]{2,}\b", text), f"{f} contains an email address"


def test_docs_cite_rule_ids_that_exist():
    if not DOCS_DIR or not (REPO / DOCS_DIR).exists():
        pytest.skip("no docs dir")
    valid = set(re.findall(r"^- ([IPCM]\d+)\.", read(REPO / "AGENTS.md"), re.M))
    assert valid
    for p in (REPO / DOCS_DIR).rglob("*"):
        if p.suffix not in (".md", ".rst"):
            continue
        for m in re.finditer(r"\brules? ([IPCM]\d+)\b", read(p), re.I):
            assert m.group(1).upper() in valid, f"{p.name}: unknown rule id {m.group(1)}"


def _strip_comments(code: str) -> str:
    """Blank out comments, leaving strings and line positions untouched.

    A regex over ``#`` cannot tell a comment from a ``#`` inside a string, and
    dropping only whole-line comments left trailing ones in, so a commented-out
    ``print(`` failed the gate that exists to find real ones. ``tokenize`` knows
    the difference.
    """
    lines = code.splitlines(keepends=True)
    try:
        tokens = list(tokenize.generate_tokens(io.StringIO(code).readline))
    except (tokenize.TokenError, SyntaxError, IndentationError):
        return re.sub(r"(?m)^\s*#.*$", "", code)
    for tok in tokens:
        if tok.type != tokenize.COMMENT:
            continue
        row, col = tok.start
        end_col = tok.end[1]
        line = lines[row - 1]
        lines[row - 1] = line[:col] + " " * (end_col - col) + line[end_col:]
    return "".join(lines)


def _code_files() -> list[tuple[str, str]]:
    return [
        (p.relative_to(SRC).as_posix(), _strip_comments(read(p)))
        for p in source_files()
        if not is_test_path(p)
    ]


def test_grep_gates_scan_files_at_all():
    # Without this the suite is green after any path move (M2).
    assert len(_code_files()) > 0


@pytest.mark.parametrize("name,pattern,exempt", GREP_GATES, ids=[g[0] for g in GREP_GATES])
def test_grep_gate(name, pattern, exempt):
    offenders = [rel for rel, code in _code_files() if not exempt(rel) and pattern.search(code)]
    assert offenders == [], f"{name}: {offenders}"

File: gap-trap/templates/gates/instruction-gate.test.ts
/**
 * Instruction gate. Reference implementation from gap-trap (vitest).
 *
 * The instruction files are read by every session, so they are checked like
 * code: every name a contract cites exists where the contract says it lives,
 * the portable core stays portable, the always-loaded files stay small, cited
 * commits exist, and no private data leaks into knowledge files. The grep
 * gates at the bottom are the contract Never clauses a text search settles.
 *
 * Every check here also asserts its own denominator: how many tokens it
 * resolved, how many files it scanned. A check that quietly measures nothing
 * reports the same green as a check that measured everything, and that is how
 * a gate dies without anyone noticing (M2).
 *
 * ADAPT the constants in the config block; the assertions stay as they are.
 */
import { describe, expect, it } from 'vitest';
import { execSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';

// ---- config ---------------------------------------------------------------
// ADAPT: repo root relative to this file, and the source tree contracts name.
const repoRoot = path.resolve(__dirname, '../../..');
const SRC_DIR = 'app/src';
const SOURCE_EXT = /\.(ts|tsx)$/;
/** Directories under SRC_DIR that hold tests or gates, skipped by the grep gates. */
const TEST_DIRS = ['__tests__', 'tests'];
/** ADAPT: project-specific tokens that must never appear in the portable core. */
const FORBIDDEN_IN_CORE = ['{{PRODUCT}}', '{{FRAMEWORK}}', SRC_DIR, 'agents/project'];
/** ADAPT: current count plus room. Raising it needs a reason in the commit message (C7). */
const WORD_BUDGET = 4000;
const MIN_CONTRACTS = 2; // the honest count; never pad
/** ADAPT: the files loaded into every session. All of them must exist. */
const ALWAYS_LOADED = ['AGENTS.md', 'AGENTS.project.md', 'CLAUDE.md'];
const KNOWLEDGE_FILES = [
  'agents/project/domain-context.md',
  'agents/project/glossary.md',
  'agents/project/out-of-scope.md',
  'agents/generic/agent-workflows.md',
];
/** ADAPT: where developer docs live, or '' to skip the rule-ID check. */
const DOCS_DIR = 'docs/developer-guide';
const DOC_EXT = /\.(rst|md)$/;
/**
 * ADAPT: one entry per contract Never clause a grep settles. `exempt` names
 * the sanctioned file(s). Comments are stripped before matching.
 */
const GREP_GATES: { name: string; pattern: RegExp; exempt?: (rel: string) => boolean }[] = [
  { name: 'Logging: no console calls outside the logger', pattern: /\bconsole\.\w/, exempt: (f) => f === 'lib/logger.ts' || f.startsWith('lib/log-file/') },
  { name: 'HTTP: no raw fetch or axios outside lib/http', pattern: /(^|[^.\w])fetch\s*\(|\baxios\b/, exempt: (f) => f.startsWith('lib/http') },
];
// ---- end config -----------------------------------------------------------

const appSrc = path.join(repoRoot, SRC_DIR);

function walk(dir: string, acc: string[] = []): string[] {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === 'node_modules') continue;
      walk(full, acc);
    } else if (SOURCE_EXT.test(entry.name)) {
      acc.push(full);
    }
  }
  return acc;
}

const read = (f: string) => fs.readFileSync(f, 'utf8');
const sourceFiles = walk(appSrc);
const isTestPath = (rel: string) => TEST_DIRS.some((d) => rel.split(path.sep).includes(d));
/** A contract symbol found only in a test file is not a symbol the code uses. */
const nonTestSourceText = sourceFiles
  .filter((f) => !isTestPath(path.relative(appSrc, f)))
  .map(read)
  .join('\n');

/** The text under a repo-relative path, file or directory, for symbol lookups. */
function textUnder(rel: string): string {
  const full = path.join(repoRoot, rel);
  if (!fs.existsSync(full)) return '';
  if (fs.statSync(full).isDirectory()) return walk(full).map(read).join('\n');
  return read(full);
}

function parseContracts(md: string): { name: string; body: string }[] {
  const section = md.split('## Architecture contracts')[1]?.split('\n## ')[0] ?? '';
  return section
    .split('\n### ')
    .slice(1)
    .map((block) => {
      const [name, ...rest] = block.split('\n');
      return { name: name.trim(), body: rest.join('\n') };
    });
}

/**
 * Fold a contract body's wrapped lines back into the field they continue.
 * Markdown prose wraps, and a `Path:` line long enough to wrap used to have
 * everything past the wrap silently dropped from the check.
 */
export function foldFields(body: string): string[] {
  const out: string[] = [];
  let open = false;
  for (const raw of body.split('\n')) {
    const line = raw.trim();
    if (!line) {
      open = false;
    } else if (/^(Owns|Path|Never|Gate):/.test(line)) {
      out.push(line);
      open = true;
    } else if (open) {
      out[out.length - 1] += ` ${line}`;
    }
  }
  return out;
}

// Remove comments and leave strings alone. A plain regex for block comments
// treats the "/*" inside a glob such as "src/**/*.ts" as the start of one and
// deletes every line up to the next "*/", carrying real violations out with
// it, after which the gate reports clean. Quoted strings end at the line
// break, as JS string literals do, so an apostrophe in JSX text costs one
// line rather than the rest of the file. Newlines survive either way.
// ponytail: no regex-literal state, so a literal ending in "\/" followed by
// "/" reads as a line comment. Add one if a repo trips it.
export function stripComments(code: string): string {
  let out = '';
  let i = 0;
  while (i < code.length) {
    const c = code[i];
    const next = code[i + 1];
    if (c === '/' && next === '/') {
      while (i < code.length && code[i] !== '\n') i += 1;
      continue;
    }
    if (c === '/' && next === '*') {
      i += 2;
      while (i < code.length && !(code[i] === '*' && code[i + 1] === '/')) {
        if (code[i] === '\n') out += '\n';
        i += 1;
      }
      i += 2;
      continue;
    }
    if (c === '"' || c === "'" || c === '`') {
      const multiline = c === '`';
      out += c;
      i += 1;
      while (i < code.length && code[i] !== c && (multiline || code[i] !== '\n')) {
        if (code[i] === '\\' && i + 1 < code.length) {
          out += code[i] + code[i + 1];
          i += 2;
          continue;
        }
        out += code[i];
        i += 1;
      }
      if (i < code.length && code[i] === c) {
        out += c;
        i += 1;
      }
      continue;
    }
    out += c;
    i += 1;
  }
  return out;
}

const backtickTokens = (line: string) => [...line.matchAll(/`([^`]+)`/g)].map((m) => m[1].replace(/\(\)$/, ''));
const escape = (s: string) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

describe('AGENTS.project.md architecture contracts', () => {
  const md = read(path.join(repoRoot, 'AGENTS.project.md'));
  const contracts = parseContracts(md);

  it(`holds at least ${MIN_CONTRACTS} contracts with all four lines`, () => {
    expect(contracts.length).toBeGreaterThanOrEqual(MIN_CONTRACTS);
    for (const c of contracts) {
      const fields = foldFields(c.body);
      for (const field of ['Owns:', 'Path:', 'Never:', 'Gate:']) {
        expect(fields.some((l) => l.startsWith(field)), `${c.name} missing ${field}`).toBe(true);
      }
    }
  });

  it('every symbol and path named in Path/Gate lines exists where the contract says', () => {
    let checkedOverall = 0;
    for (const c of contracts) {
      const fields = foldFields(c.body).filter((l) => l.startsWith('Path:') || l.startsWith('Gate:'));
      let checked = 0;
      for (const line of fields) {
        const tokens = backtickTokens(line);
        const paths = tokens.filter((t) => t.includes('/'));
        const symbols = tokens.filter((t) => !t.includes('/'));
        for (const token of paths) {
          expect(fs.existsSync(path.join(repoRoot, token)), `${c.name}: path ${token} missing`).toBe(true);
          checked += 1;
        }
        // A Path line names where the subsystem lives, so its symbols are
        // looked for there. Resolving them anywhere in the tree let a symbol
        // that had moved out of its own module keep the contract green.
        const scoped = line.startsWith('Path:') && paths.length > 0;
        const haystack = scoped ? paths.map(textUnder).join('\n') : nonTestSourceText;
        const where = scoped ? paths.join(', ') : `${SRC_DIR} (non-test files)`;
        for (const token of symbols) {
          expect(
            new RegExp(`\\b${escape(token)}\\b`).test(haystack),
            `${c.name}: symbol ${token} not found in ${where}`,
          ).toBe(true);
          checked += 1;
        }
      }
      expect(checked, `${c.name}: its Path/Gate lines name nothing the gate can check`).toBeGreaterThan(0);
      checkedOverall += checked;
    }
    expect(checkedOverall, 'no contract token was checked at all').toBeGreaterThan(0);
  });
});

describe('AGENTS.md stays portable and small', () => {
  it('contains no project-specific tokens', () => {
    const core = read(path.join(repoRoot, 'AGENTS.md')).toLowerCase();
    for (const token of FORBIDDEN_IN_CORE) {
      expect(core.includes(token.toLowerCase()), `AGENTS.md contains "${token}"`).toBe(false);
    }
  });

  it('the always-loaded files stay inside the word budget', () => {
    // Every file has to be there. Summing whatever happens to exist means
    // deleting an always-loaded file reads as coming in under budget.
    for (const f of ALWAYS_LOADED) {
      expect(fs.existsSync(path.join(repoRoot, f)), `${f} is missing; the budget would fall for free`).toBe(true);
    }
    const words = (f: string) => read(path.join(repoRoot, f)).split(/\s+/).filter(Boolean).length;
    const total = ALWAYS_LOADED.reduce((n, f) => n + words(f), 0);
    expect(total, `combined ${total} words > budget ${WORD_BUDGET}`).toBeLessThanOrEqual(WORD_BUDGET);
  });
});

describe('knowledge files stay evidence-backed and private-data-free (M5)', () => {
  it('every commit hash cited in domain-context exists in this repo', () => {
    const md = read(path.join(repoRoot, 'agents/project/domain-context.md'));
    // 7 to 40 hex chars with a digit: --oneline prints 7, and the digit keeps
    // English words made of a-f (acceded, defaced) out.
    const hashes = [...new Set([...md.matchAll(/\b[0-9a-f]{7,40}\b/g)].map((m) => m[0]).filter((h) => /\d/.test(h)))];
    for (const hash of hashes) {
      expect(
        () => execSync(`git cat-file -e ${hash}^{commit}`, { cwd: repoRoot, stdio: 'pipe' }),
        `cited commit ${hash} not found in history`,
      ).not.toThrow();
    }
  });

  it('knowledge files contain no emails or IP addresses', () => {
    for (const file of KNOWLEDGE_FILES) {
      const full = path.join(repoRoot, file);
      if (!fs.existsSync(full)) continue;
      const text = read(full);
      expect(text, `${file} contains an IP address`).not.toMatch(/\b\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}\b/);
      // The domain needs a letter TLD, or a pinned version such as
      // `vitest@1.2.3` reads as an email and the gate cries wolf.
      expect(text, `${file} contains an email address`).not.toMatch(/\b[\w.+-]+@[\w-]+\.[A-Za-z]{2,}\b/);
    }
  });
});

describe('developer docs cite rule IDs that exist', () => {
  it('every "rule <id>" reference resolves', () => {
    if (!DOCS_DIR || !fs.existsSync(path.join(repoRoot, DOCS_DIR))) return;
    const core = read(path.join(repoRoot, 'AGENTS.md'));
    const valid = new Set([...core.matchAll(/^- ([IPCM][0-9]+)\./gm)].map((m) => m[1]));
    expect(valid.size).toBeGreaterThan(0);
    for (const file of fs.readdirSync(path.join(repoRoot, DOCS_DIR)).filter((f) => DOC_EXT.test(f))) {
      const text = read(path.join(repoRoot, DOCS_DIR, file));
      for (const m of text.matchAll(/\brules? ([IPCM][0-9]+)\b/gi)) {
        expect(valid.has(m[1].toUpperCase()), `${file}: unknown rule id "${m[1]}"`).toBe(true);
      }
    }
  });
});

describe('contract Never clauses a grep can decide', () => {
  const codeFiles = sourceFiles
    .map((f) => [path.relative(appSrc, f), f] as const)
    .filter(([rel]) => !isTestPath(rel))
    .map(([rel, f]) => [rel, stripComments(read(f))] as const);

  it('scans source files at all', () => {
    // Without this the suite is green after any path move (M2).
    expect(codeFiles.length).toBeGreaterThan(5);
  });

  for (const gate of GREP_GATES) {
    it(gate.name, () => {
      const offenders = codeFiles.filter(([rel, code]) => !(gate.exempt?.(rel) ?? false) && gate.pattern.test(code)).map(([rel]) => rel);
      expect(offenders).toEqual([]);
    });
  }
});
